pkg/workceptor/kubernetes.go

```diff
@@ -378,47 +385,75 @@ func (kw *KubeUnit) KubeLoggingWithReconnect(streamWait *sync.WaitGroup, stdout
 				if kubeErr != nil {
 					kw.GetWorkceptor().nc.GetLogger().Debug("Error getting pod after reading stream: '%s'", kubeErr)
 				}
-				for _, condition := range erroredPod.Status.Conditions {
-					if condition.Type == corev1.PodReady && condition.Status == corev1.ConditionTrue {
-						podConditionReady = true
+
+				if erroredPod != nil {
+					for _, condition := range erroredPod.Status.Conditions {
+						if condition.Type == corev1.PodReady && condition.Status == corev1.ConditionTrue {
+							podConditionReady = true
+						}
 					}
-				}
 
-				if err == io.EOF && !podConditionReady {
-					if line != "" {
-						msg, _, _ := kw.ProcessLogLine(line, sinceTime, successfulWrite)
-						if msg != "" {
-							_, err = stdout.Write([]byte(msg + "\n"))
-							if err != nil {
-								*stdoutErr = fmt.Errorf("writing final line to stdout: %s", err)
-								kw.GetWorkceptor().nc.GetLogger().Error("Error writing final line to stdout: %s", err)
-
-								return
+					for _, containerStatus := range erroredPod.Status.ContainerStatuses {
+						if containerStatus.Name == WorkerContainerName {
+							switch {
+							case containerStatus.State.Running != nil:
+								break
+							case containerStatus.State.Waiting != nil:
```

Comment:
Will we ever be in a waiting state?

Reply:
I called this out in my [review](https://github.com/ansible/receptor/pull/1400#discussion_r2265040277), asking the same question - Once we're in this loop, is it even possible for the container state to be `Waiting`? . Sounds like it is possible (since our check for `Waiting` only delays for `ErrImagePull`, see link).

So it's OK to check it, since we weren't before, and the backoff is a good idea. Logging should help us understand next time if we're getting into this loop prematurely.

Reply:
Updated the other thread and checked the kubernetes source code: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1561-L1570. Basically, if the container is still in `Waiting` when we try to connect the logs, that call will fail. This logic (now deleted) would be after we already successfully connected to the log stream, which means the container must not have been `Waiting`.